cdap-proto/src/main/java/co/cask/cdap/proto/Id.java

```diff
@@ -349,4 +352,89 @@ public int hashCode() {
       return Objects.hashCode(namespace, category, name);
     }
   }
+
+
+  /**
+   * Id.Stream uniquely identifies a stream.
+   */
+  public static final class Stream {
+    private final String namespace;
+    private final String streamName;
+    private transient int hashCode;
+
+    private Stream(final String namespace, final String streamName) {
+      Preconditions.checkNotNull(namespace, "Namespace cannot be null.");
+      Preconditions.checkNotNull(streamName, "Stream name cannot be null.");
+
+      Preconditions.checkArgument(isId(namespace), "Stream namespace has an incorrect format.");
+      Preconditions.checkArgument(isId(streamName), "Stream name has an incorrect format.");
+
+      this.namespace = namespace;
+      this.streamName = streamName;
+    }
+
+    public Namespace getNamespace() {
+      return Id.Namespace.from(namespace);
+    }
+
+    public String getNamespaceId() {
+      return namespace;
+    }
+
+    public String getName() {
+      return streamName;
+    }
+
+    public static Stream from(Namespace id, String streamName) {
+      return new Stream(id.getId(), streamName);
+    }
+
+    public static Stream from(String namespaceId, String streamName) {
+      return new Stream(namespaceId, streamName);
+    }
+
+    public String toId() {
+      return String.format("%s.%s", namespace, streamName);
+    }
+
+    public static Stream fromId(String id) {
+      Iterable<String> comps = Splitter.on('.').omitEmptyStrings().split(id);
+      Preconditions.checkArgument(2 == Iterables.size(comps));
+
+      String namespace = Iterables.get(comps, 0);
+      String streamName = Iterables.get(comps, 1);
+      return from(namespace, streamName);
+    }
+
+    @Override
+    public int hashCode() {
+      if (hashCode == 0) {
```

Comment:
In practice, it runs faster. I forget where is the reference. Can find you that.